storage_controller/src/service/safekeeper_reconciler.rs

```diff
@@ -230,32 +245,42 @@ impl SafekeeperReconciler {
     }
     async fn run(&mut self) {
         loop {
-            // TODO add parallelism with semaphore here
             let req = tokio::select! {
                 req = self.rx.recv() => req,
                 _ = self.cancel.cancelled() => break,
             };
             let Some((req, req_cancel)) = req else { break };
+
+            let inner = self.inner.clone();
             if req_cancel.is_cancelled() {
                 continue;
             }
 
-            let kind = req.kind;
-            let tenant_id = req.tenant_id;
-            let timeline_id = req.timeline_id;
-            let node_id = req.safekeeper.skp.id;
-            self.reconcile_one(req, req_cancel)
-                .instrument(tracing::info_span!(
-                    "reconcile_one",
-                    ?kind,
-                    %tenant_id,
-                    ?timeline_id,
-                    %node_id,
-                ))
-                .await;
+            tokio::task::spawn(async move {
+                let kind = req.kind;
+                let tenant_id = req.tenant_id;
+                let timeline_id = req.timeline_id;
+                let node_id = req.safekeeper.skp.id;
+                inner
+                    .reconcile_one(req, req_cancel)
+                    .instrument(tracing::info_span!(
+                        "reconcile_one",
+                        ?kind,
+                        %tenant_id,
+                        ?timeline_id,
+                        %node_id,
+                    ))
+                    .await;
+            });
         }
     }
+}
+
+impl SafekeeperReconcilerInner {
     async fn reconcile_one(&self, req: ScheduleRequest, req_cancel: CancellationToken) {
+        let permit_res = self.concurrency_limiter.acquire().await;
```

Comment:
nit: I'd acquire the permit before spawning the task to avoid loads of idle tasks in a pathological scenario

Reply:
me too, and it's caused me some thinking on how to do it, but I couldn't find a way. The permit holds a reference to the concurrency limiter, so is bounded by its lifetime, and tokio's spawn function requires `'static` stuff to be passed to it. Even if you move the `inner` together with it, it doesn't work.

the alternative would be to get the permit twice, also in the "good" case. Do you think it's worth it?

Reply:
What about `acquire_owned`. That grabs an Arc to the inner semaphore.

Reply:
oh yeah that will work